### PR TITLE
rocSOLVER and hipSOLVER examples (part III)

### PR DESCRIPTION
**Pull request by Beanavil**
This pull request completes the roc/hipSOLVER batch of examples.

### New examples
- hipSOLVER
  - syevdx
  - syevj
  - syevj_batched
  - sygvj

### Renamed examples
- hipSOLVER 

  |Old name                                             | New name |
  |-----------------------------------------------------|----------|
  |`cholesky_decomposition`                             | `potrf`  |
  |`generalized_symmetric_eigenvalue_divide_and_conquer`| `sygvd`  |
  |`linear_least_squares`                               | `gels`   |
  |`qr_decomposition`                                   | `geqrf`  |
  |`symmetric_eigenvalue_divide_and_conquer`            | `syevd`  |
  |`singular_value_decomposition`                       | `gesvd`  |

### New (and final) directories order

```
├── hipSOLVER
│   ├── gels
│   ├── geqrf
│   ├── gesvd
│   ├── getrf
│   ├── potrf
│   ├── syevd
│   ├── syevdx
│   ├── syevj
│   ├── syevj_batched
│   ├── sygvd
│   └── sygvj
.
.
.
├── rocSOLVER
│   ├── getf2
│   ├── getri
│   ├── syev
│   ├── syev_batched
│   └── syev_strided_batched
```

### CODE REVIEW

**Comment by Beanavil** [comment]
cc @tfalders

File: Libraries/hipSOLVER/syevdx/README.md
Comment (tfalders) [inline Libraries/hipSOLVER/syevdx/README.md:36]: It might also be worth mentioning the complex-valued versions of the functions: Cheevdx and Zheevdx

File: Libraries/hipSOLVER/syevdx/README.md
Comment (tfalders) [inline Libraries/hipSOLVER/syevdx/README.md]: This is only the case for the cuSOLVER backend. With the rocSOLVER backend, devInfo >= 0.

File: Libraries/hipSOLVER/syevdx/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/syevdx/main.cpp]: Should there be an 'else` between these brackets?

File: Libraries/hipSOLVER/syevj/README.md
Comment (tfalders) [inline Libraries/hipSOLVER/syevj/README.md]: Typo on whether

File: Libraries/hipSOLVER/syevj/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/syevj/main.cpp]: I would suggest using lda instead of n in case someone wants to try changing lda

File: Libraries/hipSOLVER/syevj/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/syevj/main.cpp]: residue -> residual

File: Libraries/hipSOLVER/syevj/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/syevj/main.cpp]: If lda is used up above, it should be used here too. It might be worth considering giving each matrix its own leading dimension (e.g. ldv), but I leave that up to you.

File: Libraries/hipSOLVER/syevj/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/syevj/main.cpp]: If lda is used up above, it will need to be used for V here too

File: Libraries/hipSOLVER/syevj_batched/README.md
Comment (tfalders) [inline Libraries/hipSOLVER/syevj_batched/README.md]: Only for the cuSOLVER backend. With the rocSOLVER backend, devInfo >= 0.

File: Libraries/hipSOLVER/syevj_batched/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/syevj_batched/main.cpp:60]: if(batch_count < 1)?

File: Libraries/hipSOLVER/syevj_batched/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/syevj_batched/main.cpp]: Again, I recommend using lda instead of n

File: Libraries/hipSOLVER/sygvj/README.md
Comment (tfalders) [inline Libraries/hipSOLVER/sygvj/README.md]: Probably good to mention that the value for n is 3

File: Libraries/hipSOLVER/sygvj/README.md
Comment (tfalders) [inline Libraries/hipSOLVER/sygvj/README.md:40]: I'd suggest mentioning the complex versions (Chegvj and Zhegvj) as well

File: Libraries/hipSOLVER/sygvj/README.md
Comment (tfalders) [inline Libraries/hipSOLVER/sygvj/README.md]: I don't think the eigenvectors are calculated by divide-and-conquer (at least not with the rocSOLVER backend). IIRC, they are computed using the Jacobi method by applying Jacobi rotations to an initial identity matrix.

File: Libraries/hipSOLVER/sygvj/README.md
Comment (tfalders) [inline Libraries/hipSOLVER/sygvj/README.md]: It looks like you're mixing calls to the compatibility API with calls to the regular API, which is perfectly valid but it would be nice to be consistent.

File: Libraries/hipSOLVER/sygvj/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/sygvj/main.cpp]: Here's another place to use lda instead of n

File: Libraries/hipSOLVER/sygvj/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/sygvj/main.cpp]: I don't think you need to negate syevj_info here

File: Libraries/hipSOLVER/sygvj/main.cpp
Comment (tfalders) [inline Libraries/hipSOLVER/sygvj/main.cpp]: Typo on does

**Review by tfalders — COMMENTED** [review]
Looks good! Just a few comments to be addressed before I give my approval.

File: Libraries/hipSOLVER/syevdx/README.md
Comment (Beanavil) [reply Libraries/hipSOLVER/syevdx/README.md:36]: Done!

File: Libraries/hipSOLVER/syevdx/README.md
Comment (Beanavil) [reply Libraries/hipSOLVER/syevdx/README.md]: True, added this info to the README.

File: Libraries/hipSOLVER/syevdx/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevdx/main.cpp]: Originally there was an `else`, but then the following suggestion was made:
>Something like
>```c++
>if (/* some condition */) {
>  // clean and return
>}
>
>// continue program
>// cleanup
>```
>or
>```c++
>if (/* some condition */) {
>  goto cleanup_and_return;
>}
>
>// continue program
>
>cleanup_and_return:
>// cleanup
>```
>is generally preferred over:
>```c++
>if (/* inverse of some condition*/) {
>  // continue program
>}
>// cleanup
>```
>as it reduces nesting/indentation.

But after choosing Option 2 (goto approach) removing the braces resulted in a `cannot jump from this goto statement to its label` error, because then the jump bypassed the `eps`, `h_one`. `h_minus_one` and `h_zero` variables initialization. So the braces were left to prevent this error from appearing.

Maybe you have some other opinion about this design?

File: Libraries/hipSOLVER/syevj/README.md
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj/README.md]: Nice catch, fixed (also in syevj_batched).

File: Libraries/hipSOLVER/syevj/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj/main.cpp]: Fixed.

File: Libraries/hipSOLVER/syevj/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj/main.cpp]: Nice catch, fixed.

File: Libraries/hipSOLVER/syevj/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj/main.cpp]: Added the use of `lda`. Also for the size of AV here and in the `syev` example.

File: Libraries/hipSOLVER/syevj/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj/main.cpp]: Added the use of `lda` here as well.  Also for the size of VD here and in the `syev` example.

File: Libraries/hipSOLVER/syevj_batched/README.md
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj_batched/README.md]: Added this info to the README.

File: Libraries/hipSOLVER/syevj_batched/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj_batched/main.cpp:60]: It should be the same in this case, given that it's an integer. Or do you prefer it that way for some other reason?

File: Libraries/hipSOLVER/syevj_batched/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj_batched/main.cpp]: Fixed.

File: Libraries/hipSOLVER/sygvj/README.md
Comment (Beanavil) [reply Libraries/hipSOLVER/sygvj/README.md]: Changed!

File: Libraries/hipSOLVER/sygvj/README.md
Comment (Beanavil) [reply Libraries/hipSOLVER/sygvj/README.md:40]: Done!

File: Libraries/hipSOLVER/sygvj/README.md
Comment (Beanavil) [reply Libraries/hipSOLVER/sygvj/README.md]: True, must have been some copy-paste error. Fixed!

File: Libraries/hipSOLVER/sygvj/README.md
Comment (Beanavil) [reply Libraries/hipSOLVER/sygvj/README.md]: Oopsie, we shouldn't be mixing these. Should be fixed now.

File: Libraries/hipSOLVER/sygvj/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/sygvj/main.cpp]: Fixed!

File: Libraries/hipSOLVER/sygvj/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/sygvj/main.cpp]: True, fixed.

File: Libraries/hipSOLVER/sygvj/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/sygvj/main.cpp]: Nice catch, fixed!

File: Libraries/hipSOLVER/syevj_batched/main.cpp
Comment (tfalders) [reply Libraries/hipSOLVER/syevj_batched/main.cpp:60]: The if statement and the error message don't match. I don't mind leaving it as `batch_count <= 0`, as long as the error message specifies the correct condition.

File: Libraries/hipSOLVER/syevdx/main.cpp
Comment (tfalders) [reply Libraries/hipSOLVER/syevdx/main.cpp]: Since the braces have to be kept anyway, I don't think you're saving anything on indentation. Using else instead of goto would also make the code more consistent with the syevj_batched example.

File: Libraries/hipSOLVER/syevj_batched/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevj_batched/main.cpp:60]: Do you mean something like "Batch size should be greater than 0"?

File: Libraries/hipSOLVER/syevdx/main.cpp
Comment (Beanavil) [reply Libraries/hipSOLVER/syevdx/main.cpp]: Okey, should be changed now. BTW, `sygvd` was also changed so it's consistent with that approach.

File: Libraries/hipSOLVER/syevj_batched/main.cpp
Comment (tfalders) [reply Libraries/hipSOLVER/syevj_batched/main.cpp:60]: Oh! Silly me, I misinterpreted what would happen when `batch_count = 0`. My mistake. No change needed.

**Review by tfalders — APPROVED** [review]
Nice work! Approved!

**Review by Naraenda — APPROVED** [review]

**Pull request merged**
